src/MongoDB.EntityFrameworkCore/ChangeTracking/ListComparer.cs

```diff
@@ -0,0 +1,118 @@
+﻿/* Copyright 2023-present MongoDB Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Microsoft.EntityFrameworkCore.Diagnostics;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+
+namespace MongoDB.EntityFrameworkCore.ChangeTracking;
+
+internal sealed class ListComparer<TElement>(ValueComparer<TElement> elementComparer)
+    : ValueComparer<IEnumerable<TElement>>(
```

Comment:
So should this be `ValueComparer<IList<TElement>>` ?

Reply:
I wondered that too but EF 8 providers have them as IEnumerable and given the throw for non-IList there must be some paths that end up here. I'd rather preserve that exception than throw some kind of upstream error when it can't use the comparer at all.